// @dojo/framework/core/Destroyable.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../shim/Promise"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var Promise_1 = require("../shim/Promise");
    /**
     * No op function used to replace a Destroyable instance's `destroy` method, once the instance has been destroyed
     */
    function noop() {
        return Promise_1.default.resolve(false);
    }
    /**
     * No op function used to replace a Destroyable instance's `own` method, once the instance has been destroyed
     */
    function destroyed() {
        throw new Error('Call made to destroyed method');
    }
    var Destroyable = /** @class */ (function () {
        /**
         * @constructor
         */
        function Destroyable() {
            this.handles = [];
        }
        /**
         * Register handles for the instance that will be destroyed when `this.destroy` is called
         *
         * @param {Handle} handle The handle to add for the instance
         * @returns {Handle} A wrapper Handle. When the wrapper Handle's `destroy` method is invoked, the original handle is
         *                   removed from the instance, and its `destroy` method is invoked.
         */
        Destroyable.prototype.own = function (handle) {
            var _handles = this.handles;
            _handles.push(handle);
            return {
                destroy: function () {
                    _handles.splice(_handles.indexOf(handle));
                    handle.destroy();
                }
            };
        };
        /**
         * Destroys all handlers registered for the instance
         *
         * @returns {Promise<any} A Promise that resolves once all handles have been destroyed
         */
        Destroyable.prototype.destroy = function () {
            var _this = this;
            return new Promise_1.default(function (resolve) {
                _this.handles.forEach(function (handle) {
                    handle && handle.destroy && handle.destroy();
                });
                _this.destroy = noop;
                _this.own = destroyed;
                resolve(true);
            });
        };
        return Destroyable;
    }());
    exports.Destroyable = Destroyable;
    exports.default = Destroyable;
});
